### PR TITLE
Introducing rubocop

### LINKED ISSUE #463: Add linter for ruby
# Background

Thorough this project there have been a magnitude of codestyles used, and I suggest we do something to unify our code standard.

So to ensure our consistency when writing code for the backend part of the Yabs project we should look into actually enforcing the code standard that has been agreed upon (automatically).

# Proposal

Currently we do not have any linting for our backend. By implementing [Rubocop](https://github.com/rubocop-hq/rubocop) and [Rubocop-Rails](https://github.com/rubocop-hq/rubocop-rails) we can enforce our code standard and make sure we follow the Ruby on Rails [Style Guide](https://rails.rubystyle.guide/).

This would in the long run decrease the technical burden that has been and will be collected throughout this project, and increase the readability of our code.

# Implementation

In addition to adding the gems to the project and configuring Rubocop-Rails as seen in the [readme](https://github.com/rubocop-hq/rubocop-rails/blob/master/README.md) the easiest way to transition over would probably be to use the TODO option that's provided by Rubocop. This would generate a rubocop_todo file containing all the current offences, where they can be ticked off one at a time. The offences listed in the file would also not raise a exception, when running the linter for example in the CI.

```bash
$ rubocop --safe-auto-correct 
$ rubocop --auto-gen-config
```

A different way to implement this in our CI would be to use the GitHub superliner as a base, thus removing the need for a separate linter action for `ts, js` files.

Feel free to share your thoughts

_Originally posted by @te4-adrian-almetun-smeds in https://github.com/itggot-TE4/Yabs/discussions/385_

### PR DESCRIPTION
**Pull request by te4-adrian-almetun-smeds**
## ⚙️ What issue did you implement or fix?
Fixes #463

## 📜 Summary
- Added Rubocop linter
- Added Rubocop-Rails extension
- Configured workflow for linting
- Generated linter config
- Generated todo list

## 📝 Type of change

Please delete options that are not relevant.

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature or enhancement (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 This change requires a documentation update

## ✅ Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## 💩 How can this implementation be improved?

Add it to the husky pre commit hook

### CODE REVIEW

File: backend/app/controllers/api/v1/titles_controller.rb [hunk 1]
```diff
@@ -81,6 +81,7 @@ def set_title
 
     # Never trust parameters from the scary internet, only allow the white list through.
     def title_params
-      params.require(:title).permit(:name, :isbn, :cost, :title_type, :description, :authors, :cover, :page_count, :published_date, :subject_id)
+      params.require(:title).permit(:name, :isbn, :cost, :title_type, :description, :authors, :cover, :page_count, 
+                                    :published_date, :subject_id)
```
Comment (Te4-Jacob-sandstrom) [inline backend/app/controllers/api/v1/titles_controller.rb:85]: is this even a change?

File: backend/.rubocop.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+Rails/BulkChangeTable:
+  Exclude:
+    - 'db/migrate/20191216135511_add_columns_to_title.rb'
+
+Rails/ReversibleMigration:
+  Exclude:
+    - 'db/migrate/20181120100040_change_primary_key_on_users.rb'
+    - 'db/migrate/20181120114623_create_books.rb'
+    - 'db/migrate/20191216135730_drop_title_from_title.rb'
+    - 'db/migrate/20191218093038_remove_titles_id_from_review.rb'
```
Comment (Te4-Jacob-sandstrom) [inline backend/.rubocop.yml:28]: Why exclude migrations?

File: backend/app/jobs/data_fetch_job.rb [hunk 3]
```diff
@@ -7,13 +7,15 @@ class DataFetchJob < ApplicationJob
   def perform(*fetch)
     token = []
     scopes =  ['https://www.googleapis.com/auth/admin.directory.group.readonly','https://www.googleapis.com/auth/admin.directory.user.readonly']
-    client = GoogleApiClient.new(Google::Apis::AdminDirectoryV1::DirectoryService, scopes, Rails.application.secrets.google_client_secrets, 'johanneberg.laneservice@ga.ntig.se')
+    client = GoogleApiClient.new(Google::Apis::AdminDirectoryV1::DirectoryService, scopes,
+                                 Rails.application.secrets.google_client_secrets, 'johanneberg.laneservice@ga.ntig.se')
```
Comment (Te4-Jacob-sandstrom) [inline backend/app/jobs/data_fetch_job.rb:11]: why?

Comment (Te4-Jacob-sandstrom) [reply backend/app/jobs/data_fetch_job.rb:11]: is this rubocops doing?

**Review by Te4-Jacob-sandstrom — APPROVED** [review]
Looks good but I am not sure I agree with rubocops line splitting

File: backend/app/jobs/data_fetch_job.rb (hunk 3 above)
Comment (te4-adrian-almetun-smeds) [reply backend/app/jobs/data_fetch_job.rb:11]: Max line-length is at 125 chars, so separating it to a new line is required. And yes it's rubocops doing.

File: backend/.rubocop.yml (hunk 2 above)
Comment (te4-adrian-almetun-smeds) [reply backend/.rubocop.yml:28]: I don't see the point of ensuring reversible migrations for these old ones.

All future migrations are linted for reversible migrations, but I'd say we don't need it on these ones. I see **no** reason as to why they'd ever be reversed.

File: backend/app/controllers/api/v1/titles_controller.rb (hunk 1 above)
Comment (te4-adrian-almetun-smeds) [reply backend/app/controllers/api/v1/titles_controller.rb:85]: It's a new line so I suppose this is a change.

File: backend/.rubocop.yml (hunk 2 above)
Comment (Te4-Jacob-sandstrom) [reply backend/.rubocop.yml:28]: Okay that makes sense I didn't really understand what excluding them did before

**Review by te4-alexander-nylundgomes — APPROVED** [review]
Looks good!

**Review by Te4-Jacob-sandstrom — CHANGES_REQUESTED** [review]
re request review after fixing conflicts

**Review by Te4-Jacob-sandstrom — APPROVED** [review]

**Review by te4-alexander-nylundgomes — APPROVED** [review]
" I'd buy that for a dollar!"

**Pull request merged**